File: 6.栈与队列/8.前 K 个高频元素/347.前-k-个高频元素1.cpp
/*
 * @lc app=leetcode.cn id=347 lang=cpp
 *
 * [347] 前 K 个高频元素
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        auto cmp = [](pair<int, int> a, pair<int, int> b)
        {
            return a.first > b.first;
        };
        unordered_map<int, int> hash;
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> heap(cmp);
        vector<int> res;
        for (auto x : nums)
        {
            hash[x] ++;
        }
        for (auto [key, v] : hash)
        {
            heap.emplace(v, key);
            if (heap.size() > k) heap.pop();
        }
        while (heap.size()) res.push_back(heap.top().second), heap.pop();
        return res;
    }
};
// @lc code=end

File: HOT100/13.堆/347.前-k-个高频元素1.py
#
# @lc app=leetcode.cn id=347 lang=python3
#
# [347] 前 K 个高频元素
#
from mytools import *
# @lc code=start
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hash = defaultdict(int)
        max_cnt = 0
        for x in nums:
            hash[x] += 1
            max_cnt = max(max_cnt, hash[x])
        bucket = [[] for _ in range(max_cnt + 1)]
        for key, val in hash.items():
            bucket[val].append(key)
        res = []
        for i in range(max_cnt, -1, -1):
            if bucket[i]:
                for x in bucket[i]:
                    res.append(x)
            k -= len(bucket[i])
            if k <= 0: break
        return res
            
# @lc code=end

